modules/common/src/main/scala/surge/internal/utils/ExecutionContextProber.scala

```diff
@@ -0,0 +1,155 @@
+import akka.pattern.pipe
+import surge.internal.utils.ExecutionContextProberActor.Messages.Stop
+import surge.internal.utils.ExecutionContextProberActor.TimerKeys.{ SendProbesKey, TimeoutKey }
+
+import java.util.UUID
+import java.util.concurrent.TimeUnit
+import scala.concurrent.duration.{ FiniteDuration, _ }
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.language.postfixOps
+
+final case class ExecutionContextProberSettings(
+    targetEc: ExecutionContext,
+    initialDelay: FiniteDuration,
+    timeout: FiniteDuration,
+    interval: FiniteDuration,
+    numProbes: Int)
+
+object ExecutionContextProberActor {
+
+  def props(settings: ExecutionContextProberSettings): Props = Props(new ExecutionContextProberActor(settings))
+
+  object Messages {
+
+    final case object Stop
+
+    final case object SendProbes
+
+    final case object Timeout
+
+    final case object HasIssues // used by the unit test
+
+  }
+
+  object TimerKeys {
+
+    final case object SendProbesKey
+
+    final case object TimeoutKey
+
+  }
+
+  sealed trait ProbingResult
+
+  final case class Ack(id: UUID) extends ProbingResult
+
+  def noOpFuture(id: UUID)(implicit ec: ExecutionContext): Future[ProbingResult] = Future {
+    Ack(id)
+  }
+
+  val warningText = "Possible thread starvation issue"
+
+}
+
+class ExecutionContextProberActor(settings: ExecutionContextProberSettings) extends Actor with ActorLogging with Timers {
+
+  import ExecutionContextProberActor.Messages._
+  import ExecutionContextProberActor._
+
+  implicit val system: ActorSystem = context.system
+
+  var detectedIssue = false // used by the unit test
+
+  override def preStart(): Unit = {
+    timers.startSingleTimer(SendProbesKey, SendProbes, settings.initialDelay)
+  }
+
+  override def receive: Receive = {
+    case SendProbes =>
+      val id = UUID.randomUUID()
+      implicit val ec: ExecutionContext = settings.targetEc
+      (1 to settings.numProbes).foreach(_ => pipe(noOpFuture(id)).to(self))
+      timers.startSingleTimer(TimeoutKey, msg = Timeout, settings.timeout)
+      context.become(collect(count = 0, expectedId = id))
+    case HasIssues =>
+      sender() ! detectedIssue
+    case Stop =>
+      timers.cancel(SendProbesKey)
+      sender() ! Done
+      context.stop(self)
+    case _ => // ignore (the uncancelled futures we didn't collect from previous check(s) ...)
+  }
+
+  def collect(count: Int, expectedId: UUID): Receive = {
+    case Timeout =>
+      log.warning(s"One of our probes timed out. ${warningText}.")
+      // the execution context is potentially unhealthy so we want to suspend our checks for a bit (to give
+      // it time to recover in case it's just a hiccup), hence interval * 3.
+      detectedIssue = true
+      timers.startSingleTimer(SendProbesKey, msg = SendProbes, settings.interval * 3)
+      context.become(receive)
+    case Status.Failure(e) =>
+      log.error(e, s"One of our probes resulted in a failed future. ${warningText}.")
```

Comment:
It'd be good to add some details about which execution context in particular we're using to the couple of logs here. Right now It will just be the default for the actor system but I can us reusing this for other execution contexts as well.

Reply:
True, I'll add some more details

Reply:
Done, marking some resolved